applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp

```diff
@@ -371,187 +340,260 @@ namespace Kratos
+            {
+                KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Critical head undetermined at " << mCriticalHead << ", max search head reached: " << criticalHeadInfo.maxCriticalHead << std::endl;
+                mExitLoop = true;
+                break;
             }
 
-            rLogCallback(kratosLogBuffer.str().c_str());
-            Logger::RemoveOutput(p_output);
+            KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Searching at head: " << mCurrentHead << std::endl;
 
-            ResetModelParts();
-            return 0;
+            std::ostringstream currentHeadStream;
+            currentHeadStream << std::setprecision(8) << std::noshowpoint << mCurrentHead;
+            std::string currentHeadString = currentHeadStream.str();
+
+            std::string progress = "Calculating head level " + currentHeadString + "m (" + std::to_string(step) + "/" + std::to_string(maxSteps) + ")";
+            rCallBackFunctions.rReportTextualProgress(progress.data());
+            rCallBackFunctions.rReportProgress(((double)step) / ((double)maxSteps));
+
+            MainExecution(model_part, p_solving_strategy, 0.0, 1.0, 1);
+
+            auto count = std::size_t{0};
+            for (Element *element : pipeElements)
+            {
+                if (element->GetValue(PIPE_ACTIVE))
+                    count += 1;
+            }
+
+            KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Open pipe elements: " << count << std::endl;
+
+            if (count == noPipeElements)
+            {
+                HandleCriticalHeadFound(criticalHeadInfo);
+                mExitLoop = true;
+            }
+
+            GeoOutputWriter writer{gid_output_settings, mWorkingDirectory, model_part};
+            writer.WriteGiDOutput(model_part, gid_output_settings);
+
+            // Update boundary conditions for next search head.
+            if (RiverBoundary->Info() == "ApplyConstantScalarValueProcess")
+            {
+                ResetModelParts();
+                KRATOS_ERROR << "ApplyConstantScalarValueProcess process search is not implemented.";
+            }
+
+            if (RiverBoundary->Info() == "ApplyConstantHydrostaticPressureProcess")
+            {
+                mCriticalHead = currentProcess->GetReferenceCoord();
+                mCurrentHead = mCriticalHead + criticalHeadInfo.stepCriticalHead;
+                currentProcess->SetReferenceCoord(mCurrentHead);
+                step++;
+            }
+
+            if (rCallBackFunctions.rShouldCancel())
+            {
+                HandleCancellationAndReset(rCallBackFunctions.rLogCallback, p_output);
+
+                return 0;
+            }
         }
-        catch (const std::exception &exc)
+        return 0;
+    }
+
+    void KratosExecute::HandleCriticalHeadFound(const CriticalHeadInfo& criticalHeadInfo)
+    {
+        if (abs(mCurrentHead - criticalHeadInfo.minCriticalHead) < 1e-9)
```

Comment:
I always prefer to use the math functions from the `std` namespace (to ensure that we don't accidentally call the C-style variants, which don't support overloading). So:
```suggestion
        if (std::abs(mCurrentHead - criticalHeadInfo.minCriticalHead) < 1e-9)
```